listeners.py

```diff
@@ -178,6 +179,86 @@ def __del__(self):
+        self.failed_tasks = dict()
+        self.rss_results = dict()
+
+    # task name set as test name with running configuration
+    def get_task_name_conf(self, name, conf):
+        if not conf:
+            return name
+        return name + ':' + conf
+
+    def process_result(self, obj):
+        # called only once on task run initiated
+        if isinstance(obj, WorkerCurrentTask):
+            task_name = self.get_task_name_conf(obj.task_name, obj.task_param)
+            self.worker_current_task[obj.worker_id] = obj
+            self.rss_results[task_name] = get_proc_stat_rss(
+                    self.worker_id_to_pid[obj.worker_id])
+
+        # called only once on task run finished
+        if isinstance(obj, WorkerTaskResult):
+            task_name = self.get_task_name_conf(obj.task_id[0], obj.task_id[1])
+            if obj.short_status == 'fail':
+                self.failed_tasks[task_name] = 1
+
+    # called on each 'delta_seconds' delay within task run
+    def process_timeout(self, delta_seconds):
+        rss = None
+        worker_ids = self.get_not_done_worker_ids()
+
+        running_tasks = [task for worker_id, task
+                         in self.worker_current_task.items()
+                         if worker_id in worker_ids]
+        for task in running_tasks:
+            task_name = self.get_task_name_conf(task.task_name, task.task_param)
+            if task_name in self.rss_results:
+                rss = get_proc_stat_rss(self.worker_id_to_pid[task.worker_id])
+                if rss > self.rss_results[task_name]:
+                    self.rss_results[task_name] = rss
+
+    # called only once after all tasks finished
+    def print_statistics(self):
+        top_rss = 5
+        results_sorted = []
+
+        # prepare standalone subpath '<vardir>/statistics' for statistics files
+        stats_dir = os.path.join(Options().args.vardir, 'statistics')
+        safe_makedirs(stats_dir)
+
+        # RSS
+
+        # print to stdout RSS statistics for all failed tasks
+        if self.failed_tasks:
+            color_stdout('RSS of the failed tasks:\n', schema='test_var')
+            for task_name in self.failed_tasks:
+                if task_name in self.rss_results:
+                    color_stdout('* %6.1f %s\n' % (self.rss_results[task_name] / 1000, task_name),
+                                 schema='test_var')
+
+        # print to stdout RSS statistics for some number of most it used tasks
+        color_stdout('Up to {} most RSS used tasks in Mb:\n' . format(top_rss), schema='test_var')
+        results_sorted = sorted(self.rss_results.items(), key=lambda x: x[1], reverse=True)
+        count = 0
+        for task_name, rss in results_sorted:
+            if count < top_rss:
+                color_stdout('* %6.1f %s\n' % (int(rss) / 1000, task_name), schema='test_var')
+            count = count + 1
+        color_stdout('-' * 41, "\n", schema='separator')
```

Comment:
You can loop over `results_sorted[:top_rss]` instead of the `if` block and the counter.

Reply:
Right, corrected.